### DISCUSSION TITLE
Sudden, Unexpected, and Alarming Changes In Syntax

### DISCUSSION THREAD

**Discussion by Wajinn** [vlang/v#16556] [General]
First, want to say that I like the language and what team V has done so far, so any criticisms that I have, are with the best intentions in mind.  With that out of the way, here we go. The sudden unexpected changes in the syntax of Vlang can make it hard on many users and supporters, who may not be so quick to update their code. Just V progressing from alpha to beta, resulted in many changes where older code no longer works, and it can be hard enough to get around to updating it and dealing with many errors and issues. It is not to say that such changes should not be expected with an alpha or beta, but rather the suddenness and lack of open discussion about it can be unsettling.

For example, not too long ago there was the sudden change in string interpolation. It was not so much that the change was being made, but the suddenness of it and the confusion and issues that came along with it. To the extent that the change had to be reversed. I'm glad that logic prevailed, but it did have me a bit worried.

Now there is the change in generics, which is about as equally worrisome as the change with string interpolation was, and possibly as bothersome. There is logic and justifications behind why, but I can't find any open discussions about it or with any detailed comparisons. `[]` is used for slices and arrays too, and I can see where in spots it can get confusing, not to mention it can be unexpected as so many languages use `<>`. 

So, hey guys, can we talk about this a little more? What say ye all?

**Comment by JalonSolov** [comment]
There was no public discussion about the change in generics, but it was done to simplify the compiler.

Not all languages use `<>`, and the trend now is towards `[]` mainly because of the complexities parsing `<>` correctly, especially when those symbols are used for other things, like comparisons, bit-shifting, etc.

See this page for reference:  https://soc.me/languages/stop-using-angle-brackets-for-generics.html

**Reply by Wajinn** [reply]
Thanks for elaborating on this further.  Hoping that there would be more open discussion on such changes, even if its just in the context of helping to prepare people for a particular change, and it may go a long way when the logic and reasons of it are explained.

**Reply by Delta456** [reply]
All these discussions happen on the Discord Server so you will need to join the server.

**Reply by Wajinn** [reply]
> Discord Server

Yes, that's true.  And it was announced on Twitter, after the fact.  But what I'm getting at is perhaps making proposal type announcements in advance of the action  and putting it on the main site, in the changelog, as an RFC, etc...  Somewhere more public, which doesn't require signing up or logging in to anything.  For example, Go had this publicly in their "Type Parameters Proposal" (which was linked to from the Twitter announcement).

Or maybe even publicly state that Discord is where any past or future decisions have and will be made.  Possibly mentioning this on Twitter, that it was decided on Discord, when making the announcement.  As it is, there was a little (after the fact) push back on Twitter.

> [Marcin]
> Replying to 
> [@v_language](https://twitter.com/v_language)
> "C# and Kotlin tried to retain a more consistent syntax by introducing unlimited look-ahead: Their parser just keeps reading input after the < until it can make a decision." What's wrong with that approach? Isn't it shifting the complexity to the programmers? Most of us use <> :)

Just mentioning that being more obviously public, removes the perception of any changes looking sudden and unexpected, particularly as V is now in beta and more people are using, considering, or looking. It arguably looks better as an obviously debated process where the best logical conclusion was reached and any dissenting opinions were already heard, and both sides can be publically read.

Sometimes this is also good, because years after the fact people might be asking, "Why did we do that?", and the V related reasons (with the V related debate) behind the decisions can be more easily found and read.

**Comment by Wajinn** [comment]
> There was no public discussion about the change in generics, but it was done to simplify the compiler.
> See this page for reference: https://soc.me/languages/stop-using-angle-brackets-for-generics.html

There was an interesting debate (not at all saying all points were valid or good) on [Hacker news - Language design: Use [ ] instead of < > for Generics](https://news.ycombinator.com/item?id=22786049) , that possibly should have been had openly on Vlang's discussions.  There were some interesting opposing arguments to the use of `[]` for generics which were brought up.
